src/ekat/kokkos/ekat_kokkos_utils.hpp

```diff
@@ -97,6 +152,66 @@ struct ExeSpaceUtils<Kokkos::Cuda> {
   static TeamPolicy get_team_policy_force_team_size (Int ni, Int team_size) {
     return TeamPolicy(ni, team_size);
   }
+
+
+
+  template <typename TeamMember, class Lambda, typename ValueType/*, bool Serialize=true*/>
+  static KOKKOS_INLINE_FUNCTION
+  void parallel_reduce (const TeamMember& team,
+			const int& begin,
+			const int& end,
+                        const Lambda& lambda,
+                        const bool serialize,
+                        ValueType& result)
+  {
+    if (serialize)
+    {
+      // We want to get C++ on GPU to match F90 on CPU. Thus, need to
+      // serialize parallel reductions.
+
+      // All threads init result.
+      // NOTE: we *need* an automatic temporary, since we do not know
+      //       where 'result' comes from. If result itself is an automatic
+      //       variable, using it would be fine (only one vector lane would
+      //       actually have a nonzero value after the single). But if
+      //       result is taken from a view, then all vector lanes would
+      //       see the updated value before the vector_reduce call,
+      //       which will cause the final answer to be multiplied by the
+      //       size of the warp.
+      auto local_tmp = Kokkos::reduction_identity<ValueType>::sum();
+
+      Kokkos::single(Kokkos::PerThread(team),[&] {
+      for (int k=begin; k<end; ++k)
+        lambda(k, local_tmp);
+        });
+
+      // Broadcast result to all threads by doing sum of one thread's
+      // non-0 value and the rest of the 0s.
+      Kokkos::Impl::CudaTeamMember::vector_reduce(Kokkos::Sum<ValueType>(local_tmp));
+
+      result = local_tmp;
+    }
+    else
+      Kokkos::parallel_reduce(Kokkos::TeamThreadRange(team, begin, end), lambda, result);
+  }
+
+  template<typename PackType, typename ValueType, bool Serialize=false>
```

Comment:
This doesn't address the primary need. Luca suggested looking at https://github.com/E3SM-Project/E3SM/blob/b7b269b3cfaca869e861fa01e6d65e201dc588a3/components/homme/src/share/cxx/ColumnOps.hpp#L474, and I agree that's the best approach. The minor issue of providing a reduction over a single pack can wait for a future PR, and when it's done, it will live in ekat_pack.hpp, since it has nothing to do with Kokkos.

Reply:
Ok, yes I wasn't sure since the suggestion in the issue listed only the single pack version, but I will implement this tomorrow.

Reply:
Please see https://github.com/E3SM-Project/EKAT/issues/26#issuecomment-688989928 and subsequent comments.

Reply:
This is being moved to  a new PR